### PR TITLE
Tkurth/sgbn fixes

### PR DESCRIPTION
**Pull request by azrael417**
This PR fixes the single node group batch norm in APEX to work with cuda 12.2 and RTC.

### CODE REVIEW

File: apex/contrib/csrc/cudnn_gbn/norm_sample.cpp [hunk 1]
```diff
@@ -27,448 +27,453 @@
...
-        }
-    } else {
-        // Here we assume that the format is CUDNN_TENSOR_NHWC
-        strideA[1]          = 1;
-        strideA[nbDims - 1] = strideA[1] * dimA[1];
-        for (int64_t d = nbDims - 2; d >= 2; d--) {
-            strideA[d] = strideA[d + 1] * dimA[d + 1];
-        }
-        strideA[0] = strideA[2] * dimA[2];
+void generateStrides(const int64_t* dimA, int64_t* strideA, int64_t nbDims, cudnnTensorFormat_t filterFormat) {
```
Comment (eqy) [inline apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:53]: Is it possible to simply use the Tensor's existing `.strides()` rather than relying on another helper function? AFAIK it would respect the NHWC vs. NCHW convention.

File: apex/contrib/csrc/cudnn_gbn/norm_sample.cpp [hunk 2]
```diff
@@ -27,448 +27,453 @@
...
+  int64_t peer_stride[4];
+
+  // NHWC format. GenerateStrides() takes care of this. Howeever, tensor dims should still be NCHW
+  generateStrides(tensorDims, tensor_stride, (int64_t)4, CUDNN_TENSOR_NHWC);
+  generateStrides(peerDims, peer_stride, (int64_t)4, CUDNN_TENSOR_NHWC);
+
+  auto tensor_create = [&tensor_stride, &tensorDims](cudnnDataType_t type,
+  int64_t id) {
+    return cudnn_frontend::TensorBuilder()
+      .setDim(4, tensorDims)
```
Comment (eqy) [inline apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:95]: Similarly, can we use the existing `.sizes()` instead of creating another `tensorDims` array?

File: apex/contrib/csrc/cudnn_gbn/cudnn_gbn.cpp [hunk 3]
```diff
@@ -38,28 +44,43 @@ at::Tensor gbn_forward(const at::Tensor& x,
...
-    rank_id
-  );
 
+  // check if plan already exists
+  std::vector<int64_t> fv = {(int64_t)BN_FWD, N, C, H, W, bn_group, (int64_t)CUDNN_DATA_HALF};
+  if ( gbn_plan_cache.find(fv) == gbn_plan_cache.end() ) {
+    auto plan = run_batch_norm_forward(tensorDims, perChannelDims, epsilonDims, peerDims, CUDNN_DATA_HALF);
+    gbn_plan_cache.insert(std::make_pair(fv, plan));
+  }
+
```
Comment (eqy) [inline apex/contrib/csrc/cudnn_gbn/cudnn_gbn.cpp:62]: It looks like some of the code makes assumptions about the input tensor(s)' memory layout. If so, there should be checks like `is_contiguous(at::MemoryFormat::ChannelsLast)`.

File: apex/contrib/csrc/cudnn_gbn/norm_sample.cpp [hunk 4]
```diff
@@ -27,448 +27,453 @@
...
+        .setStrides(4, tensor_stride)
+          .setId(id)
+            .setAlignment(16)
```
Comment (eqy) [inline apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:98]: Manually setting alignment without checking the actual tensor address seems dangerous.

**Review by eqy — CHANGES_REQUESTED** [review]
Took a first look, but given the size of this PR I believe @crcrpar  should get the final say

**Review by crcrpar — COMMENTED** [review]
wouldn't this require any changes to https://github.com/NVIDIA/apex/blob/30a7ad3974b32f7ce68cefabc38374fb4520a35e/apex/contrib/test/cudnn_gbn/test_cudnn_gbn_with_two_gpus.py?

File: apex/contrib/csrc/cudnn_gbn/norm_sample.cpp (hunk 1 above)
Comment (azrael417) [reply apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:53]: this is how it was implemented before though, I basically based the new implementation on the old one. We could of course use strides and pass the stride tensors to the routine. Do we want to change that?

File: apex/contrib/csrc/cudnn_gbn/cudnn_gbn.cpp (hunk 3 above)
Comment (azrael417) [reply apex/contrib/csrc/cudnn_gbn/cudnn_gbn.cpp:62]: This is done on the python frontend. That check it [here](https://github.com/azrael417/apex/blob/96b961f38128cce0452d53c8b501e28e220b28e2/apex/contrib/cudnn_gbn/batch_norm.py#L122)

File: apex/contrib/csrc/cudnn_gbn/norm_sample.cpp (hunk 4 above)
Comment (azrael417) [reply apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:98]: the existing code has that too: this is just a refactor of the code which is already present:

https://github.com/NVIDIA/apex/blob/master/apex/contrib/csrc/cudnn_gbn/norm_sample.cpp

This is somewhat urgent since it is fixing a showstopper bug for mlperf hpc 3.0. I am fine with rewriting this but I want to move fast on this. Is there an example of how this should be done?

Comment (eqy) [reply apex/contrib/csrc/cudnn_gbn/norm_sample.cpp:98]: Yes, e.g., https://github.com/pytorch/pytorch/blob/004ff536e87c9586064fb49c4e581f185f3a9d47/aten/src/ATen/native/cudnn/Conv_v8.cpp#L55

**Comment by eqy** [comment]
@azrael417 we can defer addressing the issues I brought up to a later PR if @crcrpar is content to merge the fix given the urgency

**Comment by crcrpar** [comment]
rel: https://github.com/NVIDIA/apex/issues/1689
